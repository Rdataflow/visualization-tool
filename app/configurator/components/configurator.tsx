import { Trans } from "@lingui/macro";
import { SxProps } from "@mui/material";
import Box from "@mui/material/Box";
import Button, { ButtonProps } from "@mui/material/Button";
import { useRouter } from "next/router";
import React from "react";

import { SelectDatasetStep } from "@/browser/select-dataset-step";
import { ChartPanel } from "@/components/chart-panel";
import { ChartPreview } from "@/components/chart-preview";
import { PublishChartButton } from "@/components/chart-selection-tabs";
import { HEADER_HEIGHT } from "@/components/header";
import { Loading } from "@/components/hint";
import {
  ConfiguratorState,
  getChartConfig,
  useConfiguratorState,
} from "@/configurator";
import { Description, Title } from "@/configurator/components/annotators";
import { ChartAnnotationsSelector } from "@/configurator/components/chart-annotations-selector";
import { ChartConfigurator } from "@/configurator/components/chart-configurator";
import { ChartOptionsSelector } from "@/configurator/components/chart-options-selector";
import {
  ConfiguratorDrawer,
  DRAWER_WIDTH,
} from "@/configurator/components/drawer";
import { IconButton } from "@/configurator/components/icon-button";
import {
  PanelBodyWrapper,
  PanelHeaderLayout,
  PanelHeaderWrapper,
  PanelLayout,
} from "@/configurator/components/layout";
import { ChartConfiguratorTable } from "@/configurator/table/table-chart-configurator";
import SvgIcChevronLeft from "@/icons/components/IcChevronLeft";
import { useLocale } from "@/locales/use-locale";
import { useDataSourceStore } from "@/stores/data-source";
import { InteractiveFiltersProvider } from "@/stores/interactive-filters";
import useEvent from "@/utils/use-event";

const BackContainer = (props: React.PropsWithChildren<{ sx?: SxProps }>) => {
  const { children, sx } = props;

  return (
    <Box
      sx={{
        px: 2,
        minHeight: 78,
        display: "flex",
        alignItems: "center",
        ...sx,
      }}
    >
      {children}
    </Box>
  );
};

export const BackButton = ({
  children,
  onClick,
  ...props
}: { children: React.ReactNode } & ButtonProps) => {
  const { sx } = props;

  return (
    <Button
      variant="text"
      color="primary"
      size="small"
      sx={{ fontWeight: "bold", ...sx }}
      startIcon={<SvgIcChevronLeft />}
      onClick={onClick}
    >
      {children}
    </Button>
  );
};

const isAnnotationField = (field: string | undefined) => {
  return field === "title" || field === "description";
};

const useAssureCorrectDataSource = (stateGuard: ConfiguratorState["state"]) => {
  const [state] = useConfiguratorState();
  const { dataSource, setDataSource } = useDataSourceStore();

  React.useEffect(() => {
    if (state.state !== stateGuard) {
      return;
    }

    if (state.dataSource.url !== dataSource.url) {
      setDataSource(state.dataSource);
    }
  }, [
    dataSource.url,
    setDataSource,
    state.dataSource,
    state.state,
    stateGuard,
  ]);
};

const ConfigureChartStep = () => {
  const [state, dispatch] = useConfiguratorState();
  const chartConfig = getChartConfig(state);
  const router = useRouter();

  const handleClosePanel = useEvent(() => {
    dispatch({
      type: "ACTIVE_FIELD_CHANGED",
      value: undefined,
    });
  });

  const handlePrevious = useEvent(() => {
    if (state.state !== "CONFIGURING_CHART") {
      return;
    }

    router.push(
      {
        pathname: `/browse`,
        query: {
          dataset: chartConfig.cubes[0].iri,
        },
      },
      undefined,
      { shallow: true }
    );
  });

  useAssureCorrectDataSource("CONFIGURING_CHART");

  if (state.state !== "CONFIGURING_CHART") {
    return null;
  }

  return (
    <InteractiveFiltersProvider>
      <PanelLayout type="LM">
        <PanelBodyWrapper
          type="L"
          sx={{
            flexGrow: 1,
            display: "flex",
            height: "100%",
            flexDirection: "column",
          }}
        >
          <BackContainer>
            <BackButton onClick={handlePrevious}>
              <Trans id="controls.nav.back-to-preview">Back to preview</Trans>
            </BackButton>
          </BackContainer>
          {chartConfig.chartType === "table" ? (
            <ChartConfiguratorTable state={state} />
          ) : (
            <ChartConfigurator state={state} />
          )}
        </PanelBodyWrapper>
        <PanelBodyWrapper type="M">
          <ChartPanel>
            <ChartPreview dataSource={state.dataSource} />
          </ChartPanel>
        </PanelBodyWrapper>
        <ConfiguratorDrawer
          anchor="left"
          open={!!chartConfig.activeField}
          hideBackdrop
          onClose={handleClosePanel}
        >
          <div style={{ width: DRAWER_WIDTH }} data-testid="panel-drawer">
            <BackContainer>
              <Button
                variant="text"
                color="inherit"
                size="small"
                sx={{ fontWeight: "bold" }}
                startIcon={<SvgIcChevronLeft />}
                onClick={handleClosePanel}
              >
                <Trans id="controls.nav.back-to-main">Back to main</Trans>
              </Button>
            </BackContainer>
            {isAnnotationField(chartConfig.activeField) ? (
              <ChartAnnotationsSelector state={state} />
            ) : (
              <ChartOptionsSelector state={state} />
            )}
          </div>
        </ConfiguratorDrawer>
      </PanelLayout>
    </InteractiveFiltersProvider>
  );
};

const LayoutingStep = () => {
  const locale = useLocale();
  const [state, dispatch] = useConfiguratorState();
  const handlePrevious = useEvent(() => {
    if (state.state !== "LAYOUTING") {
      return;
    }
    dispatch({ type: "STEP_PREVIOUS" });
  });

  useAssureCorrectDataSource("LAYOUTING");

  if (state.state !== "LAYOUTING") {
    return null;
  }

  return (
    <InteractiveFiltersProvider>
      <PanelLayout type="LM" sx={{ background: (t) => t.palette.muted.main }}>
        <PanelHeaderLayout type="LMR">
          <PanelHeaderWrapper type="L">
            <BackContainer>
              <BackButton onClick={handlePrevious}>
                <Trans id="controls.nav.back-to-configurator">
                  Back to editing
                </Trans>
              </BackButton>
            </BackContainer>
          </PanelHeaderWrapper>
          <PanelHeaderWrapper
            type="M"
            sx={{
              display: "flex",
              alignItems: "center",
              justifyContent: "center",
              gap: 3,
              mx: "auto",
            }}
          >
            <IconButton
              label="layoutTab"
              checked={state.layout.type === "tab"}
              onClick={() => {
                if (state.layout.type === "tab") {
                  return;
                }

                dispatch({
                  type: "LAYOUT_CHANGED",
                  value: {
                    type: "tab",
                    activeField: undefined,
                  },
                });
              }}
            />
            <IconButton
              label="layoutDashboard"
              checked={state.layout.type === "dashboard"}
              onClick={() => {
                if (state.layout.type === "dashboard") {
                  return;
                }

                dispatch({
                  type: "LAYOUT_CHANGED",
                  value: {
                    type: "dashboard",
                    activeField: undefined,
                  },
                });
              }}
            />
          </PanelHeaderWrapper>
          <PanelHeaderWrapper
            type="R"
            sx={{
              display: "flex",
              alignItems: "start",
              justifyContent: "flex-end",
            }}
          >
            <PublishChartButton />
          </PanelHeaderWrapper>
        </PanelHeaderLayout>
        <PanelBodyWrapper type="L">{state.layout.type}</PanelBodyWrapper>
        <PanelBodyWrapper type="M">
          <Box
            sx={{
              display: "flex",
              flexDirection: "column",
              gap: 1,
              mt: 3,
              mb: 4,
            }}
          >
            <Title text={state.meta.title[locale]} onClick={() => {}} />
            <Description
              text={state.meta.description[locale]}
              onClick={() => {}}
            />
          </Box>
          <ChartPanel>
            <ChartPreview dataSource={state.dataSource} />
          </ChartPanel>
        </PanelBodyWrapper>
      </PanelLayout>
    </InteractiveFiltersProvider>
  );
};

const PublishStep = () => {
  const [state] = useConfiguratorState();

  if (state.state !== "PUBLISHING") {
    return null;
  }

  return (
    <PanelLayout type="LM">
      <PanelBodyWrapper type="M">
        <ChartPanel>
          <InteractiveFiltersProvider>
            <ChartPreview dataSource={state.dataSource} />
          </InteractiveFiltersProvider>
        </ChartPanel>
      </PanelBodyWrapper>
    </PanelLayout>
  );
};

export const Configurator = () => {
  const { pathname } = useRouter();
  // Local state, the dataset preview doesn't need to be persistent.
  // FIXME: for a11y, "updateDataSetPreviewIri" should also move focus to "Weiter" button (?)
  const [{ state }] = useConfiguratorState();
  const isLoadingConfigureChartStep =
    state === "INITIAL" && pathname === "/create/[chartId]";

  return isLoadingConfigureChartStep ? (
    <LoadingConfigureChartStep />
  ) : state === "SELECTING_DATASET" ? (
    <SelectDatasetStep />
  ) : (
    <>
      {state === "CONFIGURING_CHART" && <ConfigureChartStep />}
      {state === "LAYOUTING" && <LayoutingStep />}
      {state === "PUBLISHING" && <PublishStep />}
    </>
  );
};

const LoadingConfigureChartStep = () => {
  return (
    <Box
      sx={{
        display: "flex",
        alignItems: "center",
        justifyContent: "center",
        width: "100%",
        height: `calc(100vh - ${HEADER_HEIGHT}px)`,
      }}
    >
      <Loading delayMs={0} />
    </Box>
  );
};
