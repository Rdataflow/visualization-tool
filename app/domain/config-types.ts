import * as t from "io-ts";

export type ChartType = "bar" | "line" | "area" | "scatterplot" | "column";

const FilterValueMulti = t.type(
  {
    type: t.literal("multi"),
    values: t.record(t.string, t.union([t.literal(true), t.undefined])) // undefined values will be removed when serializing to JSON
  },
  "FilterValueMulti"
);
export type FilterValueMulti = t.TypeOf<typeof FilterValueMulti>;

const FilterValueSingle = t.type(
  {
    type: t.literal("single"),
    value: t.string
  },
  "FilterValueSingle"
);
export type FilterValueSingle = t.TypeOf<typeof FilterValueSingle>;

const FilterValueRange = t.type(
  {
    type: t.literal("range"),
    from: t.string,
    to: t.string
  },
  "FilterValueRange"
);
export type FilterValueRange = t.TypeOf<typeof FilterValueRange>;

const FilterValue = t.union(
  [FilterValueSingle, FilterValueMulti, FilterValueRange],
  "FilterValue"
);
export type FilterValue = t.TypeOf<typeof FilterValue>;

export type FilterValueType = FilterValue["type"];

export type FilterValueMultiValues = FilterValueMulti["values"];

const Filters = t.record(t.string, FilterValue, "Filters");

export type Filters = t.TypeOf<typeof Filters>;

const NoneConfig = t.type(
  {
    chartType: t.literal("none"),
    filters: Filters
  },
  "NoneConfig"
);
export type NoneConfig = t.TypeOf<typeof NoneConfig>;

const AreaConfig = t.type(
  {
    chartType: t.literal("area"),
    filters: Filters
  },
  "AreaConfig"
);
export type AreaConfig = t.TypeOf<typeof AreaConfig>;

const BarConfig = t.type(
  {
    chartType: t.literal("bar"),
    filters: Filters
  },
  "BarConfig"
);
const ColumnConfig = t.type(
  {
    chartType: t.literal("column"),
    filters: Filters
  },
  "BarConfig"
);
export type BarConfig = t.TypeOf<typeof BarConfig>;

const LineConfig = t.type(
  {
    chartType: t.literal("line"),
    filters: Filters
  },
  "LineConfig"
);
export type LineConfig = t.TypeOf<typeof LineConfig>;

const ScatterPlotConfig = t.type(
  {
    chartType: t.literal("scatterplot"),
    filters: Filters
  },
  "ScatterPlotConfig"
);
export type ScatterPlotConfig = t.TypeOf<typeof ScatterPlotConfig>;

// interface IriBrand {
//   readonly IRI: unique symbol;
// }
// const Iri = t.brand(
//   t.string,
//   (s): s is t.Branded<string, IriBrand> => true,
//   "IRI"
// );

const ChartConfig = t.intersection([
  t.union([
    AreaConfig,
    BarConfig,
    ColumnConfig,
    LineConfig,
    ScatterPlotConfig,
    NoneConfig
  ]),
  t.record(t.string, t.any)
]);
export type ChartConfig = t.TypeOf<typeof ChartConfig>;

const Config = t.type(
  {
    dataSet: t.string,
    // filters: Filters,
    chartConfig: ChartConfig
  },
  "Config"
);

export type Config = t.TypeOf<typeof Config>;

export const isValidConfig = (config: unknown): config is Config =>
  Config.is(config);

export const decodeConfig = (config: unknown) => Config.decode(config);

const ConfiguratorStateInitial = t.type({ state: t.literal("INITIAL") });
const ConfiguratorStateSelectingDataSet = t.type({
  state: t.literal("SELECTING_DATASET"),
  dataSet: t.undefined,
  chartConfig: NoneConfig
});

const ConfiguratorStateSelectingChartType = t.intersection([
  t.type({
    state: t.literal("SELECTING_CHART_TYPE")
  }),
  Config
]);
const ConfiguratorStateConfiguringChart = t.intersection([
  t.type({
    state: t.literal("CONFIGURING_CHART")
  }),
  Config
]);
const ConfiguratorStatePublishing = t.intersection([
  t.type({
    state: t.literal("PUBLISHING")
  }),
  Config
]);
const ConfiguratorStatePublished = t.intersection([
  t.type({
    state: t.literal("PUBLISHED"),
    configKey: t.string
  }),
  Config
]);

export type ConfiguratorStatePublishing = t.TypeOf<
  typeof ConfiguratorStatePublishing
>;

const ConfiguratorState = t.union([
  ConfiguratorStateInitial,
  ConfiguratorStateSelectingDataSet,
  ConfiguratorStateSelectingChartType,
  ConfiguratorStateConfiguringChart,
  ConfiguratorStatePublishing,
  ConfiguratorStatePublished
]);

export type ConfiguratorState = t.TypeOf<typeof ConfiguratorState>;

export const isValidConfiguratorState = (
  state: unknown
): state is ConfiguratorState => ConfiguratorState.is(state);
