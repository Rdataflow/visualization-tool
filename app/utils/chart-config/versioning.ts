import produce from "immer";

export const CHART_CONFIG_VERSION = "1.0.2";

type Migration = {
  description: string;
  from: string;
  to: string;
  up: (config: any) => any;
  down: (config: any) => any;
};

const migrations: Migration[] = [
  {
    description: `MAP
    baseLayer {
      + locked
      + bbox
    }`,
    from: "1.0.0",
    to: "1.0.1",
    up: (config: any) => {
      let newConfig = { ...config, version: "1.0.1" };

      if (newConfig.chartType === "map") {
        const { baseLayer } = newConfig;
        newConfig = produce(newConfig, (draft: any) => {
          draft.baseLayer = {
            show: baseLayer.show,
            locked: false,
          };
        });
      }

      return newConfig;
    },
    down: (config: any) => {
      let newConfig = { ...config, version: "1.0.0" };

      if (newConfig.chartType === "map") {
        const { baseLayer } = newConfig;
        newConfig = produce(newConfig, (draft: any) => {
          draft.baseLayer = {
            show: baseLayer.show,
          };
        });
      }

      return newConfig;
    },
  },
  {
    description: `MAP
    symbolLayer {
      + colors {
        + type
        + value
        + opacity
      }
      - color
    }`,
    from: "1.0.1",
    to: "1.0.2",
    up: (config: any) => {
      let newConfig = { ...config, version: "1.0.2" };

      if (newConfig.chartType === "map") {
        const { fields } = newConfig;
        const { symbolLayer } = fields;
        const { show, componentIri, measureIri, color } = symbolLayer;
        newConfig = produce(newConfig, (draft: any) => {
          draft.fields.symbolLayer = {
            show,
            componentIri,
            measureIri,
            colors: {
              type: "fixed",
              value: color,
              opacity: 80,
            },
          };
        });
      }

      return newConfig;
    },
    down: (config: any) => {
      let newConfig = { ...config, version: "1.0.1" };

      if (newConfig.chartType === "map") {
        const { fields } = newConfig;
        const { symbolLayer } = fields;
        const { show, componentIri, measureIri, colors } = symbolLayer;
        newConfig = produce(newConfig, (draft: any) => {
          draft.fields.symbolLayer = {
            show,
            componentIri,
            measureIri,
            color: colors.value,
          };
        });
      }

      return newConfig;
    },
  },
];

export const migrateChartConfig = (
  config: any,
  {
    fromVersion,
    toVersion = CHART_CONFIG_VERSION,
  }: { fromVersion?: string; toVersion?: string } = {}
): any => {
  const fromVersionFinal = fromVersion || config.version || "1.0.0";
  const direction = upOrDown(fromVersionFinal, toVersion);

  if (direction === "same") {
    return config;
  }

  const currentMigration = migrations.find(
    (migration) =>
      migration[direction === "up" ? "from" : "to"] === fromVersionFinal
  );

  if (currentMigration) {
    const newConfig = currentMigration[direction](config);
    return migrateChartConfig(newConfig, { fromVersion, toVersion });
  }

  return config;
};

const upOrDown = (
  fromVersion: string,
  toVersion: string
): "up" | "down" | "same" => {
  const fromNumbers = fromVersion.split(".").map((d) => +d);
  const toNumbers = toVersion.split(".").map((d) => +d);

  for (let i = 0; i < fromNumbers.length; i++) {
    if (fromNumbers[i] < toNumbers[i]) {
      return "up";
    }

    if (fromNumbers[i] > toNumbers[i]) {
      return "down";
    }
  }

  return "same";
};
