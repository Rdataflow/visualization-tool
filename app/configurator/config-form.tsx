import { SelectChangeEvent, SelectProps } from "@mui/material";
import get from "lodash/get";
import React, {
  ChangeEvent,
  InputHTMLAttributes,
  SyntheticEvent,
  useCallback,
  useContext,
  useMemo,
} from "react";

import { getFieldComponentIri } from "@/charts";
import { ChartConfig, ChartType } from "@/configurator/config-types";
import {
  getChartOptionBooleanField,
  getFilterValue,
  useConfiguratorState,
} from "@/configurator/configurator-state";
import { FIELD_VALUE_NONE } from "@/configurator/constants";
import {
  DimensionHierarchyQuery,
  DimensionValuesQuery,
} from "@/graphql/query-hooks";
import { DataCubeMetadata } from "@/graphql/types";
import useEvent from "@/lib/use-event";
import { CheckboxStateController, makeTreeFromValues } from "@/rdf/tree-utils";

// interface FieldProps {
//   name: HTMLInputElement["name"]
//   onChange: [];
// }
export type Option = {
  value: string | $FixMe;
  label: string | $FixMe;
  position?: number;
  isNoneValue?: boolean;
  disabled?: boolean;
};

export type FieldProps = Pick<
  InputHTMLAttributes<HTMLInputElement>,
  "onChange" | "id" | "name" | "value" | "checked" | "type"
>;

// Generic ------------------------------------------------------------------

export const useChartFieldField = ({
  field,
  dataSetMetadata,
}: {
  field: string;
  dataSetMetadata: DataCubeMetadata;
}): SelectProps => {
  const [state, dispatch] = useConfiguratorState();

  const onChange = useCallback<(e: SelectChangeEvent<unknown>) => void>(
    (e) =>
      e.target.value !== FIELD_VALUE_NONE
        ? dispatch({
            type: "CHART_FIELD_CHANGED",
            value: {
              field,
              dataSetMetadata,
              componentIri: e.target.value as string,
            },
          })
        : dispatch({
            type: "CHART_FIELD_DELETED",
            value: {
              field,
              dataSetMetadata,
            },
          }),
    [dispatch, field, dataSetMetadata]
  );

  let value: string | undefined;
  if (state.state === "CONFIGURING_CHART") {
    value =
      getFieldComponentIri(state.chartConfig.fields, field) ?? FIELD_VALUE_NONE;
  }

  return {
    name: field,
    value,
    onChange,
  };
};

export const useChartOptionSelectField = <ValueType extends {} = string>({
  field,
  path,
  getValue,
  getKey,
}: {
  field: string;
  path: string;
  getValue?: (key: string) => ValueType | undefined;
  getKey?: (value: ValueType) => string;
}): SelectProps => {
  const [state, dispatch] = useConfiguratorState();

  const onChange = useCallback<(e: SelectChangeEvent<unknown>) => void>(
    (e) => {
      const value = e.target.value as string;
      dispatch({
        type: "CHART_OPTION_CHANGED",
        value: {
          field,
          path,
          value: getValue ? getValue(value) : value,
        },
      });
    },
    [dispatch, field, path, getValue]
  );

  let value: ValueType | undefined;
  if (state.state === "CONFIGURING_CHART") {
    value = get(state, `chartConfig.fields["${field}"].${path}`);
  }
  return {
    name: path,
    value: getKey ? getKey(value!) : (value as unknown as string),
    // checked,
    onChange,
  };
};

export const useDimensionSelection = (dimensionIri: string) => {
  const [state, dispatch] = useConfiguratorState();

  const selectAll = useCallback(() => {
    dispatch({
      type: "CHART_CONFIG_FILTER_RESET_MULTI",
      value: {
        dimensionIri,
      },
    });
  }, [dispatch, dimensionIri]);

  const selectNone = useCallback(() => {
    dispatch({
      type: "CHART_CONFIG_FILTER_SET_NONE_MULTI",
      value: { dimensionIri },
    });
  }, [dispatch, dimensionIri]);

  return useMemo(() => ({ selectAll, selectNone }), [selectAll, selectNone]);
};

export const useChartOptionRadioField = ({
  field,
  path,
  value,
}: {
  field: string;
  path: string;
  value: string;
}): FieldProps => {
  const [state, dispatch] = useConfiguratorState();

  const onChange = useCallback(
    (e: ChangeEvent<HTMLInputElement>) => {
      dispatch({
        type: "CHART_OPTION_CHANGED",
        value: {
          field,
          path,
          value,
        },
      });
    },
    [dispatch, field, path, value]
  );
  const stateValue =
    state.state === "CONFIGURING_CHART"
      ? get(state, `chartConfig.fields["${field}"].${path}`, "")
      : "";
  const checked = stateValue ? stateValue === value : undefined;

  return {
    name: path,
    value,
    checked,
    onChange,
  };
};

export const useChartOptionBooleanField = ({
  path,
  field,
  defaultValue = "",
}: {
  path: string;
  field: string | null;
  defaultValue: boolean | string;
}): FieldProps => {
  const [state, dispatch] = useConfiguratorState();

  const onChange = useCallback<(e: ChangeEvent<HTMLInputElement>) => void>(
    (e) => {
      dispatch({
        type: "CHART_OPTION_CHANGED",
        value: {
          path,
          field,
          value: e.currentTarget.checked,
        },
      });
    },
    [dispatch, path, field]
  );
  const stateValue =
    state.state === "CONFIGURING_CHART"
      ? getChartOptionBooleanField(state, field, path, defaultValue)
      : defaultValue;
  const checked = stateValue ? stateValue : false;

  return {
    name: path,
    checked,
    onChange,
  };
};

export const useActiveFieldField = ({
  value,
}: {
  value: string;
}): FieldProps & {
  onClick: (x: string) => void;
} => {
  const [state, dispatch] = useConfiguratorState();

  const onClick = useCallback(() => {
    dispatch({
      type: "ACTIVE_FIELD_CHANGED",
      value,
    });
  }, [dispatch, value]);

  const checked = state.activeField === value;

  return {
    value,
    checked,
    onClick,
  };
};

// Specific ------------------------------------------------------------------
export const useChartTypeSelectorField = ({
  value,
  metaData,
}: {
  value: string;
  metaData: DataCubeMetadata;
}): FieldProps & {
  onClick: (e: SyntheticEvent<HTMLButtonElement>) => void;
} => {
  const [state, dispatch] = useConfiguratorState();
  const onClick = useCallback<(e: SyntheticEvent<HTMLButtonElement>) => void>(
    (e) => {
      const chartType = e.currentTarget.value as ChartType;

      dispatch({
        type: "CHART_TYPE_CHANGED",
        value: {
          chartType,
          dataSetMetadata: metaData,
        },
      });
    },
    [dispatch, metaData]
  );

  const stateValue =
    state.state === "CONFIGURING_CHART" || state.state === "DESCRIBING_CHART"
      ? get(state, "chartConfig.chartType")
      : "";

  const checked = stateValue === value;

  return {
    name: "chartType",
    value,
    checked,
    onClick,
  };
};

// Used in the configurator filters
export const useSingleFilterSelect = ({
  dimensionIri,
}: {
  dimensionIri: string;
}) => {
  const [state, dispatch] = useConfiguratorState();

  const onChange = useCallback<
    (e: Pick<SelectChangeEvent<unknown>, "target">) => void
  >(
    (e) => {
      dispatch({
        type: "CHART_CONFIG_FILTER_SET_SINGLE",
        value: {
          dimensionIri,
          value: (e.target.value === ""
            ? FIELD_VALUE_NONE
            : e.target.value) as string,
        },
      });
    },
    [dimensionIri, dispatch]
  );

  let value: string | undefined;
  if (state.state === "CONFIGURING_CHART") {
    value = get(
      state.chartConfig,
      ["filters", dimensionIri, "value"],
      FIELD_VALUE_NONE
    );
  }
  return {
    value,
    onChange,
  };
};

// Used in the Table Chart options
export const useSingleFilterField = ({
  dimensionIri,
  value,
}: {
  value: string;
  dimensionIri: string;
}): FieldProps => {
  const [state, dispatch] = useConfiguratorState();

  const onChange = useCallback<(e: ChangeEvent<HTMLInputElement>) => void>(
    (e) => {
      dispatch({
        type: "CHART_CONFIG_FILTER_SET_SINGLE",
        value: {
          dimensionIri,
          value: e.currentTarget.value,
        },
      });
    },
    [dispatch, dimensionIri]
  );

  const stateValue =
    state.state === "CONFIGURING_CHART"
      ? get(state.chartConfig, ["filters", dimensionIri, "value"], "")
      : "";

  const checked = stateValue === value;

  return {
    name: dimensionIri,
    value: value ? value : stateValue,
    checked,
    onChange,
  };
};

export const isMultiFilterFieldChecked = (
  chartConfig: ChartConfig,
  dimensionIri: string,
  value: string
) => {
  const filter = chartConfig.filters[dimensionIri];
  const fieldChecked =
    filter?.type === "multi" ? filter.values?.[value] ?? false : false;

  return fieldChecked || !filter;
};

const MultiFilterContext = React.createContext({
  activeKeys: new Set() as Set<string>,
  allValues: [] as string[],
  dimensionIri: undefined as string | undefined,
  colorConfigPath: undefined as string | undefined,
  checkboxController: new CheckboxStateController([], []),
});

export const useMultiFilterContext = () => {
  return useContext(MultiFilterContext);
};

type NN<T> = NonNullable<T>;
type GQLHierarchyValue = NN<
  NN<
    NN<DimensionHierarchyQuery["dataCubeByIri"]>["dimensionByIri"]
  >["hierarchy"]
>[number];

export const MultiFilterContextProvider = ({
  dimensionIri,
  colorConfigPath,
  dimensionData,
  hierarchyData,
  children,
}: {
  dimensionIri: string;
  hierarchyData: GQLHierarchyValue[] | undefined;
  dimensionData: NonNullable<
    DimensionValuesQuery["dataCubeByIri"]
  >["dimensionByIri"];
  children: React.ReactNode;
  colorConfigPath?: string;
}) => {
  const [state] = useConfiguratorState();

  const activeFilter = dimensionData
    ? getFilterValue(state, dimensionData.iri)
    : null;

  const allValues = useMemo(() => {
    return dimensionData?.values.map((d) => d.value) ?? [];
  }, [dimensionData?.values]);

  const activeKeys: Set<string> = useMemo(() => {
    if (!dimensionData) {
      return new Set();
    }
    const activeKeys = activeFilter
      ? activeFilter.type === "single"
        ? [String(activeFilter.value)]
        : activeFilter.type === "multi"
        ? Object.keys(activeFilter.values)
        : []
      : allValues;
    return new Set(activeKeys);
  }, [dimensionData, activeFilter, allValues]);

  const ctx = useMemo(() => {
    const tree =
      hierarchyData && hierarchyData.length > 0
        ? hierarchyData
        : makeTreeFromValues(allValues, dimensionIri, { depth: 0 });

    return {
      allValues,
      activeKeys,
      dimensionIri,
      colorConfigPath,
      checkboxController: new CheckboxStateController(tree, [...activeKeys]),
    };
  }, [hierarchyData, allValues, activeKeys, dimensionIri, colorConfigPath]);

  return (
    <MultiFilterContext.Provider value={ctx}>
      {children}
    </MultiFilterContext.Provider>
  );
};

export const useMultiFilterCheckboxes = (
  value: string,
  onChangeProp?: () => void
) => {
  const [, dispatch] = useConfiguratorState();
  const { dimensionIri, checkboxController } = useMultiFilterContext();

  const onChange = useEvent((e: ChangeEvent<HTMLInputElement>) => {
    if (!dimensionIri) {
      return;
    }
    checkboxController.toggle(value);
    dispatch({
      type: "CHART_CONFIG_FILTER_SET_MULTI",
      value: {
        dimensionIri,
        values: checkboxController.getValues(),
      },
    });
    onChangeProp?.();
  });

  const checkboxState = checkboxController.checkboxStates.get(value);
  return {
    onChange,
    checked: checkboxState === "checked",
    indeterminate: checkboxState === "indeterminate",
    dimensionIri,
  };
};

export const useMetaField = ({
  metaKey,
  locale,
  value,
}: {
  metaKey: string;
  locale: string;
  value?: string;
}): FieldProps => {
  const [, dispatch] = useConfiguratorState();

  const onChange = useCallback<(e: ChangeEvent<HTMLInputElement>) => void>(
    (e) => {
      dispatch({
        type: "CHART_DESCRIPTION_CHANGED",
        value: {
          path: `${metaKey}.${locale}`,
          value: e.currentTarget.value,
        },
      });
    },
    [dispatch, metaKey, locale]
  );

  return {
    name: `${metaKey}-${locale}`,
    value,
    onChange,
  };
};
