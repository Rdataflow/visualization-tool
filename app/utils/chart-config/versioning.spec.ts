import { decodeChartConfig } from "@/configurator/config-types";

import { migrateChartConfig } from "./versioning";

describe("config migrations", () => {
  const oldConfig = {
    version: "1.0.0",
    chartType: "map",
    interactiveFiltersConfig: {
      legend: {
        active: false,
        componentIri: "",
      },
      time: {
        active: false,
        componentIri: "",
        presets: {
          type: "range",
          from: "",
          to: "",
        },
      },
      dataFilters: {
        active: false,
        componentIris: [],
      },
    },
    filters: {},
    fields: {
      areaLayer: {
        show: true,
        componentIri: "GeoShapesDimensionIri",
        measureIri: "MeasureIri",
        colorScaleType: "continuous",
        colorScaleInterpolationType: "linear",
        palette: "oranges",
        nbClass: 5,
      },
      symbolLayer: {
        show: true,
        componentIri: "GeoCoordinatesDimensionIri",
        measureIri: "MeasureIri",
        color: "blue",
      },
    },
    baseLayer: {
      show: true,
    },
  };

  it("should migrate to newest config", () => {
    const migratedConfig = migrateChartConfig(oldConfig);
    const decodedConfig = decodeChartConfig(migratedConfig);

    expect(decodedConfig).toMatchSnapshot();
  });

  it("should migrate to initial config from migrated config", () => {
    const migratedConfig = migrateChartConfig(oldConfig);
    const migratedOldConfig = migrateChartConfig(migratedConfig, {
      toVersion: "1.0.0",
    });

    expect(migratedOldConfig).toEqual(oldConfig);
  });
});
