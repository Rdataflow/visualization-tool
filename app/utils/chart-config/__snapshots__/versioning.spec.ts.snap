// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`config migrations should migrate to newest config 1`] = `
Object {
  "baseLayer": Object {
    "bbox": undefined,
    "locked": false,
    "show": true,
  },
  "chartType": "map",
  "fields": Object {
    "areaLayer": Object {
      "colorScaleInterpolationType": "linear",
      "colorScaleType": "continuous",
      "componentIri": "GeoShapesDimensionIri",
      "measureIri": "MeasureIri",
      "nbClass": 5,
      "palette": "oranges",
      "show": true,
    },
    "symbolLayer": Object {
      "colors": Object {
        "opacity": 80,
        "type": "fixed",
        "value": "blue",
      },
      "componentIri": "GeoCoordinatesDimensionIri",
      "measureIri": "MeasureIri",
      "show": true,
    },
  },
  "filters": Object {},
  "interactiveFiltersConfig": Object {
    "dataFilters": Object {
      "active": false,
      "componentIris": Array [],
    },
    "legend": Object {
      "active": false,
      "componentIri": "",
    },
    "time": Object {
      "active": false,
      "componentIri": "",
      "presets": Object {
        "from": "",
        "to": "",
        "type": "range",
      },
    },
  },
  "version": "1.0.2",
}
`;
